### ISSUE TITLE
Hidden hydro contraint limits pumping power for small reservoirs

### ISSUE THREAD

**Issue by flomnes** [AntaresSimulatorTeam/Antares_Simulator#243]
**Description**
When using a small hydro reservoir, the pumping can be very limited.

**How to Reproduce**
**Please post study**

**Expected behavior**
To be discussed

**OS and version**
 - OS: All
 - Antares Simulator version: 8.0.2

**Comment by flomnes** [comment]
The Windows installer can be retrieved [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/suites/2595443482/artifacts/56724403).

**Comment by flomnes** [comment]
**src/solver/optimisation/opt_restaurer_les_donnees.cpp**
```cpp
PmaxHydEcretee = CaracteristiquesHydrauliques->CntEnergieH2OParJour[Jour]; 		
PmaxHydEcretee *= ProblemeHebdo->CoefficientEcretementPMaxHydraulique[Pays];  			
PmaxHydEcretee /= (double) ProblemeHebdo->NombreDePasDeTempsDUneJournee; 
// CODE ADDED BY M. DOQUET
if (CaracteristiquesHydrauliques->PresenceDePompageModulable == OUI_ANTARES)
{
	PmaxHydUplift  = CaracteristiquesHydrauliques->ContrainteDePmaxPompageHoraire[Pdt]; 
	PmaxHydUplift *= ProblemeHebdo->CoefficientEcretementPMaxHydraulique[Pays];
	// (Florian) PmaxHydUplift = min(PmaxHydUplift, CaracteristiquesHydrauliques->NiveauInitialReservoir /double(168))
	if (PmaxHydUplift*double(168) > CaracteristiquesHydrauliques->NiveauInitialReservoir)
	{
		PmaxHydUplift = CaracteristiquesHydrauliques->NiveauInitialReservoir /double(168); 
	}
	// (Florian) PmaxHydEcretee = max(PmaxHydUplift, PmaxHydEcretee)
	if (PmaxHydEcretee < PmaxHydUplift) 
		PmaxHydEcretee = PmaxHydUplift;
}
// END CODE ADDED BY M. DOQUET
// (Florian) ContrainteDePmaxHydrauliqueHoraire[Pdt] = min(PmaxHydEcretee, ContrainteDePmaxHydrauliqueHoraire[Pdt])
if (PmaxHydEcretee
    < CaracteristiquesHydrauliques->ContrainteDePmaxHydrauliqueHoraire[Pdt])
{
	CaracteristiquesHydrauliques->ContrainteDePmaxHydrauliqueHoraire[Pdt]
	= PmaxHydEcretee;
}

```

To me, it seems that PmaxHydEcretee is **increased** if the initial reservoir level is low (see comments). @PPlessiez can you confirm ?

In pseudo-code, 
```
PmaxHydUplift = min(
CntEnergieH2OParJour[Jour]*CoefficientEcretementPMaxHydraulique[Pays]/NombreDePasDeTempsDUneJournee, CaracteristiquesHydrauliques->NiveauInitialReservoir /double(168))

PmaxHydEcretee = max(PmaxHydUplift, PmaxHydEcretee)

ContrainteDePmaxHydrauliqueHoraire[Pdt] = min(PmaxHydEcretee, ContrainteDePmaxHydrauliqueHoraire[Pdt])
```

**Comment by PPlessiez** [comment]
Thanks for the code!

If I sum things up according to how I understand the code (and this ends up being the same things as the pseudo-code you wrote, but with additional steps so I can understand it):

PmaxHydEcretee = CntEnergieH2OParJour[Jour] * CoefficientEcretementPMaxHydraulique[Pays] / 24 _: it seems that CntEnergieH2OParJour[Jour] is the daily energy credit coming from the hydro pre-allocation and CoefficientEcretementPMaxHydraulique[Pays] is the "intra-daily modulation" parameter available in the GUI, can you confirm?_

If pumping is available on the reservoir {
PmaxHydUpLift = ContrainteDePmaxPompageHoraire[Pdt] * CoefficientEcretementPMaxHydraulique[Pays]
PmaxHydUplift = min(PmaxHydUplift, NiveauInitialReservoir /double(168))
PmaxHydEcretee = max(PmaxHydUplift, PmaxHydEcretee)
}

ContrainteDePmaxHydrauliqueHoraire[Pdt] = min(PmaxHydEcretee, ContrainteDePmaxHydrauliqueHoraire[Pdt])

From the code, it seems that when the reservoir level is low, PmaxHydUpLift is low, and has no influence on PmaxHydEcretee, I agree with you. But this is definitely not what we observe as Antares behaviour, so I'm a bit puzzled here. Can you also confirm that the final output ContrainteDePmaxHydrauliqueHoraire[Pdt] is what will be applied as the max generating hydro power in the weekly optimization problem?

**Comment by PPlessiez** [comment]
Ok, I think I got it: the thing is that the issue arises for a reservoir with very low or no inflows (e.g. PSP with no natural inflows). In that case, in the first step, PmaxHydEcretee = 0 (no daily energy credit). Then PmaxHydEcretee is increased up to PmaxHydUpLift, but as the reservoir is small, PmaxHydUpLift is low as well, and the final PmaxHydrauliqueHoraire is really small.

In that case, maybe it would make sense to just set PmaxHydEcretee to infinite in case there are no inflows in the reservoir, to avoid this side effect.
@JMJ-rte what do you think?

**Comment by JMJ-rte** [comment]
Ok, all this seems to be unclear and the formula "PmaxHydUplift = min(PmaxHydUplift, NiveauInitialReservoir /double(168))" is strange as it links constraints on flow to a specific absolute level of reservoir.
It seems that we should have additionnal constraint to avoid negative volumes of reservoir that the ones we have to avoid volume of reservoir lower than lower RC.
We have to keep in mine that usually the "0" level of reservoir is not a fully empty reservoir but the sort of minimum level of water the reservoir management should reach. There is no reason that have specific additional constraint related to this level and mathematically for Antares problem, I do not see any problem if this level sometimes is negative.
For sure the "exagerate" constraint added with this formula is an issue, as for PSP small close loop a very high reservoir volume and possibility to move only between 98 to 100% of this volume is required to eliminate its undesired effects.

So my suggestion in order to minimize colateral effects is just to comment this formula :
 # "PmaxHydUplift = min(PmaxHydUplift, NiveauInitialReservoir /double(168))"

**Comment by PPlessiez** [comment]
We need to be careful on this, as on a reservoir with no inflows, we still need to have a PmaxHydUpLift to have some possible generation (as PmaxHydEcretee will be equal to 0 in that case).

If we just remove this formula, then PmaxHydUpLift would be equal to ContrainteDePmaxPompageHoraire * CoefficientEcretementPMaxHydraulique (which is the intra-daily modulation, I think), and PmaxHydEcretee equal to the same value.

To sum things up, for a reservoir with no inflows:
- With my proposal to set PmaxHydEcretee to infinite, then ContrainteDePmaxHydrauliqueHoraire[Pdt] = ContrainteDePmaxHydrauliqueHoraire[Pdt] (the max generating power set by the user in the GUI).
- With Jean-Marc proposal to remove the line including the initial reservoir level, then PmaxHydEcretee = ContrainteDePmaxPompageHoraire[Pdt] * CoefficientEcretementPMaxHydraulique[Pays], and ContrainteDePmaxHydrauliqueHoraire[Pdt] = min(PmaxHydEcretee, ContrainteDePmaxHydrauliqueHoraire[Pdt]) (so the max power output can be limited to this "Pumping power * Intra-daily modulation" value).

**Comment by PPlessiez** [comment]
Results of the first tests:
- On a case with a small closed-loop PSP (AT closed-loop PSP from MAF 20), a simulation without this constraint in V8.0.2 gives results which are identical to a case where the reservoir capacity of this PSP has been artificially increased in V8.0.2 (with the corresponding rule curves): this is the expected behaviour.
- On a study with no explicit modelling of pumping (BP 19), a simulation without this constraint in V8.0.2 gives results which are identical to a simulation with the constraint included in V8.0.2: this is the expected behaviour.
- On a study with one node having explicit modelling of pumping (BP19 + explicit modelling of pumping for the FR weekly PSP), a simulation without this constraint in V8.0.2 gives results which are identical to a simulation with the constraint included in V8.0.2: this is the expected behaviour.

**Comment by PPlessiez** [comment]
Last test performed:

- On a standard ERAA case with no explicit modelling of pumping, a simulation without this constraint in V8.1 gives identical results to a simulation with the constraint included in V8.1.1: this is the expected behaviour.
